lib/upsert.rb

```diff
@@ -37,6 +37,23 @@ def logger
       end
     end
 
+    def mutex_for_row(upsert, row)
+      retrieve_mutex(upsert.table_name, row.selector.keys)
+    end
+
+    def mutex_for_function(upsert, row)
+      retrieve_mutex(upsert.table_name, row.selector.keys, row.setter.keys)
+    end
+
+    def retrieve_mutex(*args)
+      # ||= isn't an atomic operation
+      MUTEX_FOR_PERFORM.synchronize do
+        @mutex_cache ||= {}
+      end
+      cache_key = args.flatten.join('::')
+      @mutex_cache[cache_key] ||= Mutex.new
```

Comment:
this probably also has to be inside the mutex

Reply:
Good point.